LuaUI/Widgets/Shaders/DrawPrimitiveAtUnit.vert.glsl

```diff
@@ -0,0 +1,107 @@
+#version 420
+#extension GL_ARB_uniform_buffer_object : require
+#extension GL_ARB_shader_storage_buffer_object : require
+#extension GL_ARB_shading_language_420pack: require
+// This shader is (c) Beherith (mysterme@gmail.com)
+
+#line 5000
+
+layout (location = 0) in vec4 lengthwidthcornerheight;
+layout (location = 1) in uint teamID;
+layout (location = 2) in uint numvertices;
+layout (location = 3) in vec4 parameters; // time start, animate, unused, unused
+layout (location = 4) in vec4 uvoffsets; // this is optional, for using an Atlas
+layout (location = 5) in uvec4 instData;
+
+//__ENGINEUNIFORMBUFFERDEFS__
+//__DEFINES__
+
+struct SUniformsBuffer {
+    uint composite; //     u8 drawFlag; u8 unused1; u16 id;
+    
+    uint unused2;
+    uint unused3;
+    uint unused4;
+
+    float maxHealth;
+    float health;
+    float unused5;
+    float unused6;
+    
+    vec4 drawPos;
+    vec4 speed;
+    vec4[4] userDefined; //can't use float[16] because float in arrays occupies 4 * float space
+};
+
+layout(std140, binding=1) readonly buffer UniformsBuffer {
+    SUniformsBuffer uni[];
+}; 
+
+#define UNITID (uni[instData.y].composite >> 16)
+
+#line 10000
+
+uniform float addRadius = 0.0;
+uniform float iconDistance = 20000.0;
+
+out DataVS {
+	uint v_numvertices;
+	float v_rotationY;
+	vec4 v_color;
+	vec4 v_lengthwidthcornerheight;
+	vec4 v_centerpos;
+	vec4 v_uvoffsets;
+	vec4 v_parameters;
+	#if (FULL_ROTATION == 1)
+		mat3 v_fullrotation;
+	#endif
+};
+
+layout(std140, binding=0) readonly buffer MatrixBuffer {
+	mat4 UnitPieces[];
+};
+
+
+bool vertexClipped(vec4 clipspace, float tolerance) {
+  return any(lessThan(clipspace.xyz, -clipspace.www * tolerance)) ||
+         any(greaterThan(clipspace.xyz, clipspace.www * tolerance));
+}
+
+void main()
+{
+	uint baseIndex = instData.x; // this tells us which unit matrix to find
+	mat4 modelMatrix = UnitPieces[baseIndex]; // This gives us the models  world pos and rot matrix
+
+	gl_Position = cameraViewProj * vec4(modelMatrix[3].xyz, 1.0); // We transform this vertex into the center of the model
+	v_rotationY = atan(modelMatrix[0][2], modelMatrix[0][0]); // we can get the euler Y rot of the model from the model matrix
+	v_uvoffsets = uvoffsets;
+	v_parameters = parameters;
+	// teamID 255 is used for local selection color
+	v_color = teamID == 255 ? vec4(0.1, 1.0, 0.2, 1.0) : teamColor[teamID];
+	v_centerpos = vec4( modelMatrix[3].xyz, 1.0); // We are going to pass the centerpoint to the GS
+	v_lengthwidthcornerheight = lengthwidthcornerheight;
+	#if (ANIMATION == 1)
+	    // No animation when parameters.y is 0
```

Comment:
Also my addition. I wanted to use the same shader for local selections and ally selections.